File: clojure/twelve-days/src/twelve_days.clj
(ns twelve-days (:require [clojure.string :as str]))

(def ordinals {1 "first" 2 "second" 3 "third" 4 "fourth" 5 "fifth" 6 "sixth"
               7 "seventh" 8 "eighth" 9 "ninth" 10 "tenth" 11 "eleventh" 12 "twelfth"})

(def gifts {2 "two Turtle Doves,"
            3 "three French Hens,"
            4 "four Calling Birds,"
            5 "five Gold Rings,"
            6 "six Geese-a-Laying,"
            7 "seven Swans-a-Swimming,"
            8 "eight Maids-a-Milking,"
            9 "nine Ladies Dancing,"
            10 "ten Lords-a-Leaping,"
            11 "eleven Pipers Piping,"
            12 "twelve Drummers Drumming,"})

(defn- recite-verse
  "Returns the lyrics of a specified verse."
  [verse]
  (str "On the " (ordinals verse) " day of Christmas my true love gave to me: "
       (let [a-partridge "a Partridge in a Pear Tree."]
         (if (= verse 1) a-partridge 
           (str (str/join " " (map gifts (range verse 1 -1))) " and " a-partridge)))))

(defn recite
  "Returns the lyrics of the song: 'The Twelve Days of Christmas.'"
  [start-verse end-verse]
  (str/join "\n" (map recite-verse (range start-verse (inc end-verse)))))

File: python/twelve-days/twelve_days.py
"""Return the lyrics of the song The Twelve Days of Christmas"""

GIFTS = (
    '',
    'a Partridge in a Pear Tree',
    'two Turtle Doves',
    'three French Hens',
    'four Calling Birds',
    'five Gold Rings',
    'six Geese-a-Laying',
    'seven Swans-a-Swimming',
    'eight Maids-a-Milking',
    'nine Ladies Dancing',
    'ten Lords-a-Leaping',
    'eleven Pipers Piping',
    'twelve Drummers Drumming'
)

ORDINALS = (
    '', 'first', 'second', 'third', 'fourth', 'fifth', 'sixth',
    'seventh', 'eighth', 'ninth', 'tenth', 'eleventh', 'twelfth'
)

def recite_verse(verse_num):
    gifts = [GIFTS[index] for index in range(verse_num, 0, -1)]
    if verse_num > 1:
        gifts[-1] = f'and {gifts[-1]}'

    return (
        f'On the {ORDINALS[verse_num]} day of Christmas my true love gave to me: '
        f'{", ".join(gifts)}.'
    )

    
def recite(start_verse, end_verse):
    return [recite_verse(num) for num in range(start_verse, end_verse + 1)]
